sdk/core/azure_core/src/http/policies/instrumentation/request_instrumentation.rs

```diff
@@ -0,0 +1,483 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+use super::{
+    AZ_CLIENT_REQUEST_ID_ATTRIBUTE, AZ_NAMESPACE_ATTRIBUTE, AZ_SERVICE_REQUEST_ID_ATTRIBUTE,
+    ERROR_TYPE_ATTRIBUTE, HTTP_REQUEST_METHOD_ATTRIBUTE, HTTP_REQUEST_RESEND_COUNT_ATTRIBUTE,
+    HTTP_RESPONSE_STATUS_CODE_ATTRIBUTE, SERVER_ADDRESS_ATTRIBUTE, SERVER_PORT_ATTRIBUTE,
+    URL_FULL_ATTRIBUTE,
+};
+use crate::{
+    http::{headers, Context, Request},
+    tracing::{Span, SpanKind},
+};
+use std::sync::Arc;
+use typespec_client_core::{
+    http::policies::{Policy, PolicyResult, RetryPolicyCount},
+    tracing::Attribute,
+};
+
+/// Sets distributed tracing information for HTTP requests.
+#[derive(Clone, Debug)]
+pub(crate) struct RequestInstrumentationPolicy {
+    tracer: Option<Arc<dyn crate::tracing::Tracer>>,
+}
+
+impl RequestInstrumentationPolicy {
+    /// Creates a new `RequestInstrumentationPolicy`.
+    ///
+    /// # Arguments
+    /// - `tracer`: Pre-configured tracer to use for instrumentation.
+    ///
+    /// # Returns
+    /// A new instance of `RequestInstrumentationPolicy`.
+    ///
+    /// # Note
+    ///
+    /// The tracer provided is a "fallback" tracer which is used if the `ctx` parameter
+    /// to the `send` method does not contain a tracer.
+    ///
+    pub fn new(tracer: Option<Arc<dyn crate::tracing::Tracer>>) -> Self {
+        Self { tracer }
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+impl Policy for RequestInstrumentationPolicy {
+    async fn send(
+        &self,
+        ctx: &Context,
+        request: &mut Request,
+        next: &[Arc<dyn Policy>],
+    ) -> PolicyResult {
+        // If the context has a tracer (which happens when called from an instrumented method),
+        // we prefer the tracer from the context.
+        // Otherwise, we use the tracer from the policy itself.
+        // This allows for flexibility in using different tracers in different contexts.
+
+        // We use `.or_else` here instead of `.or` because `.or` eagerly evaluates the right-hand side,
+        // which can lead to unnecessary overhead if the tracer is not needed.
+        #[allow(clippy::unnecessary_lazy_evaluations)]
+        let tracer = ctx
+            .value::<Arc<dyn crate::tracing::Tracer>>()
+            .or_else(|| self.tracer.as_ref());
+
+        // If there is a span in the context, if it's not recording, just forward the request
+        // without instrumentation.
+        if let Some(span) = ctx.value::<Arc<dyn Span>>() {
+            if !span.is_recording() {
```

Comment:
same as one of the prev comments, we can't short-circuit here, we need to keep span active and inject context into headers regardless of whether it's being recorded to correlate other telemetry.